File: codechef/START83C/5_ROTMIN.py
from math import ceil

# https://www.codechef.com/problems/ROTMIN

# worked in cpp, not in python
    
t = int(input())
for _ in range(t):
    n, p, q = map(int, input().split())
    s = input()
    positive = []
    negative = []
    for i in s:
        if i == 'a':
            positive.append(0)
        else:
            positive.append(123  - ord(i))
        negative.append(ord(i) - 97)

    beg = -1
    end = n
    possible = 0
    p_left = p
    q_left = q

    while ( beg != end):
        mid = ceil((beg + end)/2)
        pos = 0
        neg = 0
        positive1 = positive[:mid+1]
        negative1 = negative[:mid+1]
        positive1.sort()
        negative1.sort()
        pi=0
        qi=0
        while (pos<=p and neg<=q and pi+qi<mid+1):
            if pos + positive1[pi] <= p and neg + negative1[qi] <= q:
                if positive1[pi] <= negative1[qi]:
                    pos += positive1[pi]
                    pi += 1
                else:
                    neg += negative1[qi]
                    qi += 1
            elif pos + positive1[pi] <= p:
                pos += positive1[pi]
                pi += 1
            elif neg + negative1[qi] <= q:
                neg += negative1[qi]
                qi += 1
            else:
                break
        if pi + qi < mid+1:
            end = mid-1
        else:
            beg = mid+1
            possible = mid
            p_left = p - pos
            q_left = q - neg

    p = p_left
    q = q_left
    # print("till ",possible, " position we can have 'a'")
    # print("remaining p,q = ",p_left," ", q_left)
    focus = s[possible:]
    pos = 0
    neg = 0
    # print("string on focus: ", focus)
    for i in range(len(focus)):
        if focus[i] == 'a':
            continue
        else:
            pos1 = 123 - ord(focus[i])
            neg1 = ord(focus[i]) - 97
            if pos + pos1 <= p:
                pos += pos1
                focus = focus[:i]  + 'a' + focus[i+1:]
            elif neg + neg1 <= q:
                neg += neg1
                focus = focus[:i]  + 'a' + focus[i+1:]
            elif neg < q:
                focus = focus[:i]  + chr(ord(focus[i]) - (q - neg)) + focus[i+1:]
                neg = q
            elif pos < p:
                continue
            else: 
                break
    print('a' * (possible) + focus)
        

File: codechef/START83C/5_ROTMIN.cpp
#include<bits/stdc++.h>

#define ios_base ::synch_with_stdio(false); cin.tie(0); cout.tie(0);

using namespace std;



string solve(string s, int p, int q){
    int pos = 0;
    int neg = 0;
    for (int i=0; i<s.size(); i++)
        if (s[i] == 'a') continue;
        else{
            int pos1 = 123 - s[i];
            int neg1 = s[i] - 97;
            // if ((pos + pos1 <= p) && (neg + neg1 <= q) && (i+1 < s.size())){
                // s[i] = 'a';
                // s = s.substr(0,i+1) + min(solve(s.substr(i+1), p-pos-pos1, q-neg), solve(s.substr(i+1), p-pos, q-neg-neg1));
                // break;
            // }
            if (pos + pos1 <= p){
                s[i] = 'a';
                pos += pos1;
            }
            else if ((neg + neg1 <= q)){
                s[i] = 'a';
                neg += neg1;
            }
            else if (neg < q){
                s[i] = s[i] - (q-neg);
                neg = q;
            }
            else if (pos < p) continue;
            else break;
        }
    return s;
}


int main(){

    int t;
    // freopen("5_ROTMIN_input.txt", "r", stdin);
    cin>>t;
    while(t--){
        int n, p, q;
        cin>>n>>p>>q;
        string s;
        cin>>s;
        // cout<<s<<endl;
        int positive[n], negative[n];
        for (int i=0;i<n;i++){
            // if (s[i] == 'a') positive[i] = 0;
            positive[i] = 123 - s[i];
            negative[i] = s[i] - 97;
        }
        int beg = -1;
        int end = n;

        int pleft = p;
        int qleft = q;
        int possible = 0;

        while(beg+1!=end){
            // cout<<"begin "<<beg<<" end "<<end<<"\n";
            int mid = (beg + end)/2;
            vector<int> positive1 = vector<int>(positive, positive + mid+1);
            sort(positive1.begin(), positive1.end());
            vector<int> negative1 = vector<int>(negative, negative + mid+1);
            sort(negative1.begin(), negative1.end());
            int pi = 0;
            int qi = 0;
            int pos = 0;
            int neg = 0;
            // cout<<"positive1: ";
            // for(auto x: positive1) cout<<x<<" ";
            // cout<<"\nnegative1: ";
            // for(auto x: negative1) cout<<x<<" ";
            // cout<<"\n";
            while (pos<=p && neg<=q && pi+qi<mid+1){
                // cout<<"pi: "<<pi<<" qi: "<<qi<<endl;
                if (pos + positive1[pi] <= p && neg + negative1[qi] <= q){
                    if (positive1[pi] <= negative1[qi]) pos += positive1[pi++];
                    else neg += negative1[qi++];
                }
                else if (pos + positive1[pi]<=p) pos += positive1[pi++];
                else if (neg + negative1[qi]<=q) neg += negative1[qi++];
                else break;
            }
            if (pi + qi < mid + 1)   end = mid;
            else{
                beg = mid;
                possible = mid+1;
                pleft = p-pos;
                qleft = q-neg;
            }
        }
        // cout<<"'a' is possible only till position "<<possible<<" after that only p_left: "<<pleft<<" qleft: "<<qleft<<"\n";
        
        cout<<string(possible, 'a') + solve(s.substr(possible), pleft, qleft)<<"\n";
    }
    return 0;
}